esp8266/modespwifi.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015 Paul Sokolovsky
+ * Copyright (c) 2015 Josef Gajdusek
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include "py/nlr.h"
+#include "py/obj.h"
+#include "py/gc.h"
+#include "py/runtime.h"
+#include MICROPY_HAL_H
+#include "netutils.h"
+#include "queue.h"
+#include "user_interface.h"
+#include "espconn.h"
+#include "ip_addr.h"
+#include "spi_flash.h"
+#include "utils.h"
+
+#define MODESP_INCLUDE_CONSTANTS (1)
+
+static void error_check(bool status, const char *msg) {
+    if (!status) {
+        nlr_raise(mp_obj_new_exception_msg(&mp_type_OSError, msg));
+    }
+}
+
+STATIC void esp_wifi_scan_cb(scaninfo *si, STATUS status) {
+    //printf("in pyb_scan_cb: %d, si=%p, si->pbss=%p\n", status, si, si->pbss);
+    struct bss_info *bs;
+    if (si->pbss) {
+        STAILQ_FOREACH(bs, si->pbss, next) {
+            mp_obj_tuple_t *t = mp_obj_new_tuple(6, NULL);
+            t->items[0] = mp_obj_new_bytes(bs->ssid, strlen((char*)bs->ssid));
+            t->items[1] = mp_obj_new_bytes(bs->bssid, sizeof(bs->bssid));
+            t->items[2] = MP_OBJ_NEW_SMALL_INT(bs->channel);
+            t->items[3] = MP_OBJ_NEW_SMALL_INT(bs->rssi);
+            t->items[4] = MP_OBJ_NEW_SMALL_INT(bs->authmode);
+            t->items[5] = MP_OBJ_NEW_SMALL_INT(bs->is_hidden);
+            call_function_1_protected(MP_STATE_PORT(scan_cb_obj), t);
+        }
+    }
+}
+
+STATIC mp_obj_t esp_wifi_scan(mp_obj_t cb_in) {
+    MP_STATE_PORT(scan_cb_obj) = cb_in;
+    wifi_set_opmode(STATION_MODE);
+    wifi_station_scan(NULL, (scan_done_cb_t)esp_wifi_scan_cb);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp_wifi_scan_obj, esp_wifi_scan);
+
+STATIC mp_obj_t esp_wifi_ip(mp_uint_t n_args, const mp_obj_t *args) {
```

Comment:
> @danicampora would you consider merging config_ip into ifconfig, such that if no args are given to ifconfig then it queries the config, otherwise it sets it? This would save you a few bytes.

Yes of course, if you think that makes more sense to use ifconfig for both things, then I'm happy with it :-).